MetaMorpheus/EngineLayer/Util/DictionaryExtensions.cs

```diff
@@ -0,0 +1,119 @@
+﻿using Nett;
+using Newtonsoft.Json.Linq;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Numerics;
+
+namespace EngineLayer;
+
+public static class DictionaryExtensions
+{
+    private static readonly object AddOrCreateLock = new object();
+
+    /// <summary>
+    /// Adds a value to the list associated with the specified key in the dictionary.
+    /// If the key does not exist, a new list is created with the value and added to the dictionary.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValues">The type of the values in the lists.</typeparam>
+    /// <param name="dictionary">The dictionary to operate on.</param>
+    /// <param name="key">The key whose value list to add to or create.</param>
+    /// <param name="value">The value to add to the list associated with the specified key.</param>
+    /// <remarks>
+    /// This is not thread-safe!
+    /// </remarks>
+    public static void AddOrCreate<TKey, TValues>(this IDictionary<TKey, IList<TValues>> dictionary, TKey key, TValues value)
+    {
+        if (dictionary.TryGetValue(key, out IList<TValues> values))
+        {
+            values.Add(value);
+        }
+        else
+        {
+            dictionary.Add(key, new List<TValues> { value });
+        }
+    }
+
+    /// <summary>
+    /// Adds a value to the list associated with the specified key in the dictionary.
+    /// If the key does not exist, a new list is created with the value and added to the dictionary.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValues">The type of the values in the lists.</typeparam>
+    /// <param name="dictionary">The dictionary to operate on.</param>
+    /// <param name="key">The key whose value list to add to or create.</param>
+    /// <param name="value">The value to add to the list associated with the specified key.</param>
+    /// <remarks>
+    /// This is thread safe for all dictionary types. 
+    /// </remarks>
+    public static void AddOrCreateThreadSafe<TKey, TValues>(this IDictionary<TKey, IList<TValues>> dictionary, TKey key, TValues value)
+    {
+        if (dictionary is ConcurrentDictionary<TKey, IList<TValues>> concurrentDictionary)
+        {
+            concurrentDictionary.AddOrUpdate(key, new List<TValues> { value }, (k, v) =>
+            {
+                lock (AddOrCreateLock)
+                {
```

Comment:
I don't think you should call the lock inside the addOrUpdate method. That should be handled internally by the AddOrUpdate method